spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/VectorSearchDelegate.java

```diff
@@ -0,0 +1,397 @@
+			ParameterBindingContext context) {
+
+		Integer numCandidates = null;
+		Limit limit;
+		Class<?> outputType = typeToRead != null ? typeToRead : processor.getReturnedType().getReturnedType();
+		VectorSearchInput query = queryFactory.createQuery(accessor, codec, context);
+
+		if (this.limitExpression != null) {
+			Object value = evaluator.evaluate(this.limitExpression);
+			limit = value instanceof Limit l ? l : Limit.of(((Number) value).intValue());
+		} else if (this.limit.isLimited()) {
+			limit = this.limit;
+		} else {
+			limit = accessor.getLimit();
+		}
+
+		if (limit.isLimited()) {
+			query.query().limit(limit);
+		}
+
+		if (this.numCandidatesExpression != null) {
+			numCandidates = ((Number) evaluator.evaluate(this.numCandidatesExpression)).intValue();
+		} else if (this.numCandidates != null) {
+			numCandidates = this.numCandidates;
+		} else if (query.query().isLimited() && searchType == VectorSearchOperation.SearchType.ANN) {
+
+			/*
+			MongoDB: We recommend that you specify a number at least 20 times higher than the number of documents to return (limit) to increase accuracy.
+			 */
+			numCandidates = query.query().getLimit() * 20;
+		}
+
+		return new QueryMetadata(query.path, "__score__", query.query, searchType, outputType, numCandidates,
+				getSimilarityFunction(accessor));
+	}
+
+	public String getQueryString() {
+		return queryFactory.getQueryString();
+	}
+
+	ScoringFunction getSimilarityFunction(MongoParameterAccessor accessor) {
+
+		Score score = accessor.getScore();
+
+		if (score != null) {
+			return score.getFunction();
+		}
+
+		Range<Score> scoreRange = accessor.getScoreRange();
+
+		if (scoreRange != null) {
+			if (scoreRange.getUpperBound().isBounded()) {
+				return scoreRange.getUpperBound().getValue().get().getFunction();
+			}
+
+			if (scoreRange.getLowerBound().isBounded()) {
+				return scoreRange.getLowerBound().getValue().get().getFunction();
+			}
+		}
+
+		return ScoringFunction.unspecified();
+	}
+
+	/**
+	 * Metadata for a Vector Search Aggregation.
+	 *
+	 * @param path
+	 * @param query
+	 * @param searchType
+	 * @param outputType
+	 * @param numCandidates
+	 * @param scoringFunction
+	 */
+	public record QueryMetadata(String path, String scoreField, Query query, VectorSearchOperation.SearchType searchType,
+			Class<?> outputType, @org.jspecify.annotations.Nullable Integer numCandidates, ScoringFunction scoringFunction) {
+
+		/**
+		 * Create the Aggregation Pipeline.
+		 *
+		 * @param queryMethod
+		 * @param accessor
+		 * @return
+		 */
+		public List<AggregationOperation> getAggregationPipeline(MongoQueryMethod queryMethod,
+				MongoParameterAccessor accessor) {
+
+			Vector vector = accessor.getVector();
+			Score score = accessor.getScore();
+			Range<Score> distance = accessor.getScoreRange();
+			int limit;
+
+			if (query.isLimited()) {
+				limit = query.getLimit();
+			} else {
+				limit = Math.max(1, numCandidates() != null ? numCandidates() / 20 : 1);
```

Comment:
limit cannot exceed `numCandidates` and `numCandidates` should be at least 20x higher than limit but is deriving the limit the way it's done here valid?

Reply:
Probably doesn't make sense.